[inductor] Prevent blowup in inner_fn_str and extract_read_writes

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88933

Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in #pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by formatting the string as a series of assignments to
variables, so for the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

Which corresponds to sequence of `ops` calls made.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

peterbell10 committed:
[inductor] Prevent blowup in inner_fn_str and extract_read_writes

Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in #pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by generating a basic expression tree that
exactly represents the series of `ops` calls made, with intermediate
results given explicit names. Since the old method would ignore dead
code I've also implemented a quick dead code elimination pass which
runs at the end.

For the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Prevent blowup in inner_fn_str and extract_read_writes"


Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by generating a basic expression tree that
exactly represents the series of `ops` calls made, with intermediate
results given explicit names. Since the old method would ignore dead
code I've also implemented a quick dead code elimination pass which
runs at the end.

For the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Prevent blowup in inner_fn_str and extract_read_writes"


Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by generating a basic expression tree that
exactly represents the series of `ops` calls made, with intermediate
results given explicit names. Since the old method would ignore dead
code I've also implemented a quick dead code elimination pass which
runs at the end.

For the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/dependencies.py:
```diff
@@ -185,22 +175,25 @@ def canonicalize(
     def load(self, name: str, index: sympy.Expr) -> str:
```
The type hint of the return value is `KernelBuilder.Expr` now? Same for other methods in the class?

jgong5 reviewed (commented):
We have buffer realization heuristics on CPU that relies on the `inner_fn_str`. Shall we adjust that as well?
https://github.com/pytorch/pytorch/blob/8371bb8a3dddbead709bc1e9d26715818a34fa8a/torch/_inductor/ir.py#L3808

peterbell10 commented:
The heuristic you linked shouldn't be effected by this:
https://github.com/pytorch/pytorch/blob/8371bb8a3dddbead709bc1e9d26715818a34fa8a/torch/_inductor/ir.py#L3809

On the other hand the generic heuristic is most likely improved by this change, 
https://github.com/pytorch/pytorch/blob/8371bb8a3dddbead709bc1e9d26715818a34fa8a/torch/_inductor/ir.py#L3816

jgong5 commented:
> The heuristic you linked shouldn't be effected by this

The heuristics relies on the fact that the `inner_fn_str` is an expression with the `exp` as the last op by checking "exp(" is the leading str but the PR changes the `inner_fn_str` to a sequence of assignment/return statements so checking the leading "exp(" cannot work any longer. Do I understand this correctly or I missed something here?

peterbell10 commented:
Oh I mentally substituted `startswith` to `in`. Why does it have to be the most recent op in the chain? For example `sigmoid` uses `ops.exp` but would never be realized by that heuristic because it does extra work with the result?

jgong5 commented:
> Oh I mentally substituted `startswith` to `in`. Why does it have to be the most recent op in the chain? For example `sigmoid` uses `ops.exp` but would never be realized by that heuristic because it does extra work with the result.

Sure, `in` would work fine here. Thanks for the confirmation. The original heuristics assumed that the ops would be decomposed into those with `exp` so checking the leading `exp` would suffice. Thanks for pointing out that it is not true for `sigmoid` (not sure why we don't decompose it further though). ;-)

peterbell10 committed:
Change CPU heuristic to use  on "[inductor] Prevent blowup in inner_fn_str and extract_read_writes"


Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by generating a basic expression tree that
exactly represents the series of `ops` calls made, with intermediate
results given explicit names. Since the old method would ignore dead
code I've also implemented a quick dead code elimination pass which
runs at the end.

For the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Rebase on "[inductor] Prevent blowup in inner_fn_str and extract_read_writes"


Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by generating a basic expression tree that
exactly represents the series of `ops` calls made, with intermediate
results given explicit names. Since the old method would ignore dead
code I've also implemented a quick dead code elimination pass which
runs at the end.

For the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Prevent blowup in inner_fn_str and extract_read_writes"


Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by generating a basic expression tree that
exactly represents the series of `ops` calls made, with intermediate
results given explicit names. Since the old method would ignore dead
code I've also implemented a quick dead code elimination pass which
runs at the end.

For the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Prevent blowup in inner_fn_str and extract_read_writes"


Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by generating a basic expression tree that
exactly represents the series of `ops` calls made, with intermediate
results given explicit names. Since the old method would ignore dead
code I've also implemented a quick dead code elimination pass which
runs at the end.

For the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Prevent blowup in inner_fn_str and extract_read_writes"


Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by generating a basic expression tree that
exactly represents the series of `ops` calls made, with intermediate
results given explicit names. Since the old method would ignore dead
code I've also implemented a quick dead code elimination pass which
runs at the end.

For the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented on test/inductor/test_torchinductor.py:
```diff
@@ -4954,7 +4954,7 @@ def fn(x1, x2):
                 compiled = compile_fx_inner(traced, [x1, x2])
                 assert same(fn(x1, x2)[0], compiled([x1, x2])[0], equal_nan=True)
-                assert metrics.generated_cpp_vec_kernel_count == 1
+                assert metrics.generated_cpp_vec_kernel_count == 0
```
Previously `pow(x, 10)` causes the `realize_bytes_threshold` to be exceeded and intermediate results get realized and stored to memory. This intermediate is contiguous which allows the rest of the calculations to be vectorized. However, now the entire operation gets fused into a single loop which also happens to prevent the vectorization. So while the looks like a regression, it's just the product of enabling more fusions.

jgong5 replied:
I guess this makes sense. But I'm copying @EikanWang to double confirm.

EikanWang replied:
`pow(x, 10)` is lowered as multiple multiplications. Hence, it leads to exceeding the `realize_bytes_threshold` for this case, and then the codegen will produce multiple kernels that some could be vectorized while some could not.

I go through this PR. With this PR, the codegen will only produce a single kernel for this case. @peterbell10 , am I correct?

peterbell10 replied:
Yes that's right. 

To expand further, `x` will be represented as a really long string like `div(mul(sub(sqrt(exp(cos(relu(sigmoid(square(neg(sin(abs(load(...`
and even though `pow` uses an algorithm to turn exponentiation by n into O(log(n)) multiplications, the default handler would expand that to write out all 10 multiplications as if by
```python
" * ".join([x] * 10)
```
Meaning the already long string grows over 10 times longer. However, with this PR `pow(x, 10)` causes `inner_fn_str()` to grow by a fixed number of lines, instead of multiplying the length of the string.

jansel commented on torch/_inductor/virtualized.py:
```diff
@@ -91,6 +93,102 @@ def inner(*args):
             setattr(cls, name, make_handler(format_string))
 
 
+class KernelBuilder:
```
I'd suggest renaming this to KernelPrinter, since I believe it is only used in printing not in codegen.  (Note extract_read_writes just ignores the return value.)

In codegen we have a separate solution for this:
https://github.com/pytorch/pytorch/blob/ed41a7fb681ff51129763213042da0be4fbd1a56/torch/_inductor/codegen/common.py#L432-L488
and 
https://github.com/pytorch/pytorch/blob/ed41a7fb681ff51129763213042da0be4fbd1a56/torch/_inductor/codegen/common.py#L568-L612

Overall, I'd like to simplify this given it is just for printing code.  Perhaps we could reuse some of the code used in codgen?

A more minimal version might look something like:
```py
class AssignVars:
    def __init__(self, parent_handler):
       self.parent_handler = parent_handler
       self.output = IndentedBuffer()
       self.var_counter = itertools.count()

    def __getattr__(self, name):
        def inner(*args, **kwargs):
            line = getattr(self.parent_handler, name)(*args, **kwargs)
            if name == "indirect_indexing":
                return line
            # replace line with a new variable name
            varname = f"tmp{next(self.var_counter)}"
            self.output.writeline(f"{varname} = {line}")
            return varname
       
        return inner
```

peterbell10 committed:
Remove dead code elimination and expression trees on "[inductor] Prevent blowup in inner_fn_str and extract_read_writes"


Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in #pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by formatting the string as a series of assignments to
variables, so for the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

Which corresponds to sequence of `ops` calls made.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Prevent blowup in inner_fn_str and extract_read_writes"


Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in #pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by formatting the string as a series of assignments to
variables, so for the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

Which corresponds to sequence of `ops` calls made.

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/common.py:
```diff
@@ -120,86 +125,12 @@ def remainder(a, b):
-
-
-class DeferredLine:
+class DeferredLine(DeferredLineBase):
```
This is needed to resolve an import cycle. `codegen.common` imports `virtualized`, so I had to move `IndentedBuffer`. `IndentedBuffer` needs `DeferredLine` to do `isinstance(line, DeferredLine)` and `DeferredLine.__call__` needs `virtualized.graph`. So, I've added `DeferredLineBase` which doesn't use `V.graph` but covers all the interface needed by `IndentedBuffer`.

peterbell10 commented on torch/_inductor/dependencies.py:
```diff
@@ -165,24 +164,15 @@ def merge(self, other):
-    # and cause process hanging. The result is not used.
-    # https://github.com/pytorch/torchdynamo/issues/1352
-    @staticmethod
-    def truncate_expr(expr):
```
I've removed all the `truncate_expr` code as this fixes the issue it was here to solve in a more general way.

peterbell10 commented on torch/_inductor/virtualized.py:
```diff
@@ -91,6 +90,29 @@ def inner(*args):
+
+    def getvalue(self, result):
+        self.output.writeline(f"return {result}")
+        return self.output.getvalue()
```
Without the expression tree it's no longer possible to do dead-code elimination.

peterbell10 reviewed (commented):
@jansel PTAL

peterbell10 commented:
@jansel ping

jansel commented:
@peterbell10 sorry I missed this, if you re-request review next time it will show up in my review queue.

peterbell10 committed:
Update on "[inductor] Prevent blowup in inner_fn_str and extract_read_writes"


Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in #pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by formatting the string as a series of assignments to
variables, so for the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

Which corresponds to sequence of `ops` calls made.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 committed:
Update on "[inductor] Prevent blowup in inner_fn_str and extract_read_writes"


Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in #pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by formatting the string as a series of assignments to
variables, so for the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

Which corresponds to sequence of `ops` calls made.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge -f "ROCm timeout is unrelated"

peterbell10 committed:
Fix merge conflict on "[inductor] Prevent blowup in inner_fn_str and extract_read_writes"


Currently the default `ops` handler expects strings as arguments and
just formats them into a function call template string. For complex
expressions, this can lead to exponential growth in terms. Say for
example you have:

```python
def fn(a):
   for _ in range(3)
       a = ops.mul(a, a)
   return a
```

You might expect `inner_fn_str` to contain 1 load and 3 multiplies,
but instead you find 8 loads and 7 multiplies:
```python
load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0) * load(arg_0, i0)
```

This type of blowup is present in the lowering for
`max_pool2d_with_indices_backward` which in #pytorch/torchdynamo#1352
was reported to have caused the entire compilation to hang.

This PR fixes the issue by formatting the string as a series of assignments to
variables, so for the example above, we now get:
```
tmp0 = load(arg_0, i0)
tmp1 = tmp0 * tmp0
tmp2 = tmp1 * tmp1
tmp3 = tmp2 * tmp2
return tmp3
```

Which corresponds to sequence of `ops` calls made.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
